source/particle/interpolator/bilinear_least_squares.cc

```diff
@@ -95,93 +101,172 @@ namespace aspect
-                                          c[property_index][4] * relative_support_point_location[0] * relative_support_point_location[1] +
-                                          c[property_index][5] * relative_support_point_location[0] * relative_support_point_location[2] +
-                                          c[property_index][6] * relative_support_point_location[1] * relative_support_point_location[2] +
-                                          c[property_index][7] * relative_support_point_location[0] * relative_support_point_location[1] * relative_support_point_location[2];
-                  }
-
-                // Overshoot and undershoot correction of interpolated particle property.
-                if (use_global_min_max_limiter)
+                std::size_t positions_index = 0;
+                for (typename std::vector<Point<dim>>::const_iterator itr = positions.begin(); itr != positions.end(); ++itr, ++positions_index)
                   {
-                    interpolated_value = std::min(interpolated_value, global_maximum_particle_properties[property_index]);
-                    interpolated_value = std::max(interpolated_value, global_minimum_particle_properties[property_index]);
+                    Point<dim> relative_support_point_location = mapping.transform_real_to_unit_cell(found_cell, *itr);
+                    double interpolated_value = c[property_index][0];
+                    for (unsigned int i = 1; i < n_matrix_columns; ++i)
+                      {
+                        relative_support_point_location[i - 1] -= unit_offset;
+                        interpolated_value += c[property_index][i] * relative_support_point_location[i - 1];
+                      }
+                    if (use_linear_least_squares_limiter[property_index] == true)
+                      {
+#ifdef DEBUG
+                        double init = interpolated_value;
+#endif
+                        // Due to floating point inaccuracies init and interpolated_value can differ resulting in
+                        // an overshoot or undershoot of around machine error. We resolve these small overshoot/undershoots by chopping.
+                        interpolated_value = std::min(interpolated_value, property_bounds[property_index].second);
+                        interpolated_value = std::max(interpolated_value, property_bounds[property_index].first);
+                        // If the limiter is working correctly, we should not be chopping anything more than machine error,
+                        // So the check ensures that we have chopped no more than a 1e-12th of the initial value.
+                        // To prevent division by zero, but still provide some amount of scale to our error checking,
+                        // we divide by the larger of 1 and the interpolated value
+                        Assert(std::abs(init - interpolated_value)/std::max(std::abs(interpolated_value), 1.0) <= 1e-12, ExcInternalError());
```

Comment:
I do not understand where the 1.0 comes from, does that assume the value is bounded by 1.0? I understand your concern about floating point precision, but if you write the assert in the following way you can save all the calculations above:
```suggestion
                        Assert(interpolated_value < property_bounds[property_index].second * (1.0 + std::numeric_limits<double>::epsilon()), ExcInternalError());
                        Assert(interpolated_value > property_bounds[property_index].first * (1.0 - std::numeric_limits<double>::epsilon()), ExcInternalError());
```